client/types/packagessyncer.go

```diff
@@ -0,0 +1,56 @@
+package types
+
+import (
+	"context"
+	"io"
+)
+
+// PackagesSyncer can be used by the agent to initiate syncing a package from the server.
+// The PackagesSyncer instance knows the right context: the particular OpAMPClient and
+// the particular PackageAvailable message the OnPackageAvailable callback was called for.
+type PackagesSyncer interface {
+	// Sync the available package from the server to the agent.
+	// The agent must supply an PackagesStateProvider to let the Sync function
+	// know what is available locally, what data needs to be sync and how the
+	// data can be stored locally.
+	Sync(ctx context.Context, localState PackagesStateProvider) error
+}
+
+type PackagesStateProvider interface {
+	AllPackagesHash() ([]byte, error)
+
+	// Packages returns the names of all packages that exist in the agent's local storage.
+	Packages() ([]string, error)
+
+	// PackageHash returns the hash of a local package. packageName is one of the names
+	// that were returned by Packages().
+	PackageHash(packageName string) ([]byte, error)
+
+	// CreatePackage creates the package locally. If the package existed must return an error.
+	// If the package did not exist its hash should be set to nil.
+	CreatePackage(packageName string) error
+
+	// FileContentHash returns the content hash of the package file that exists locally.
+	FileContentHash(packageName string) ([]byte, error)
+
+	// UpdateContent must create or update the package content file. The entire content
+	// of the file must be replaced by the data. The data must be read until
+	// it returns an EOF. If reading from data fails UpdateContent must abort and return
+	// an error.
+	// Content hash must be updated if the data is updated without failure.
+	// The function must cancel and return an error if the context is cancelled.
+	UpdateContent(ctx context.Context, packageName string, data io.Reader, contentHash []byte) error
+
+	// SetPackageHash must remember the hash for the specified package. Must be returned
+	// later when PackageHash is called. SetPackageHash is called after all UpsertFile
+	// and DeleteFile calls complete successfully.
```

Comment:
I don't see the referenced functions (`UpsertFile` and `DeleteFile`). Should it refer to `UpdateContent` and `DeletePackage` instead?

Reply:
Good catch, this was remaining from the time when addons had multiple files. It should be "after Update Content".